File: server/models/event_log.py
from pydantic import BaseModel, Field
from typing import Optional, Dict
from datetime import datetime

class EventLog(BaseModel):
    id: Optional[str] = Field(None, alias="_id")
    guest_id: str  # UUID for guest tracking
    event_type: str = Field(regex="^(page_view|add_to_cart|remove_from_cart|begin_checkout|purchase|add_to_wishlist|remove_from_wishlist)$")
    metadata: Optional[Dict[str, str]]  # Additional event metadata
    ip_address: Optional[str]
    created_at: Optional[datetime] = None

    class Config:
        allow_population_by_field_name = True
        schema_extra = {
            "example": {
                "guest_id": "guest123",
                "event_type": "page_view",
                "metadata": {
                    "productId": "prod123",
                    "cartValue": 100.0
                },
                "ip_address": "192.168.1.1",
                "created_at": "2026-03-10T12:00:00Z"
            }
        }

File: server/models/EventLog.js
const mongoose = require('mongoose');

const eventLogSchema = new mongoose.Schema({
  guestId: { type: String, required: true, index: true }, // UUID for guest tracking
  eventType: {
    type: String,
    enum: ['page_view', 'add_to_cart', 'remove_from_cart', 'begin_checkout', 'purchase', 'add_to_wishlist', 'remove_from_wishlist'],
    required: true,
    index: true,
  },
  metadata: {
    productId: String,
    orderId: String,
    cartValue: Number,
    itemCount: Number,
    sessionId: String,
    userAgent: String,
  },
  ipAddress: String,
  createdAt: { type: Date, default: Date.now }
});

// TTL index to automatically delete old events after 90 days (for privacy)
eventLogSchema.index({ createdAt: 1 }, { expireAfterSeconds: 90 * 24 * 60 * 60 });

// Compound index for efficient queries (guestId + eventType + createdAt)
eventLogSchema.index({ guestId: 1, eventType: 1, createdAt: -1 });

const EventLog = mongoose.model('EventLog', eventLogSchema);

module.exports = EventLog;

